### PR TITLE
Feature/doc service

### PR DESCRIPTION
**Pull request by medgar-nitric**

### CODE REVIEW

**Comment by tjholm** [comment]
Noticed the tests aren't running for this PR. But can't see any changes to the test workflow in this PR either.

File: src/main/java/io/nitric/api/document/SubCollection.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+     * @param id   the document unique id (required)
+     * @param type the document value type (required)
+     * @return new collection document reference
+     */
+    @Override
+    public <K> SubCollectionDocRef<K> doc(String id, Class<K> type) {
+        Contracts.requireNonBlank(id, "id");
+
+        if (collection.parent.id.isBlank()) {
+            throw newMissingParentException(id);
```
Comment (tjholm) [inline src/main/java/io/nitric/api/document/SubCollection.java:73]: Is this just to subcollection queries without specific parent document scope? It seems a bit fragile to implement a full interface but only allow half its functionality under specific circumstances, and the change in behaviour isn't necessarily immediately apparent to a consumer as well.

I think a less fragile design would be to be able to set the scope within a query object itself at a higher level. Would be good to discuss further. In firestore `CollectionGroups` are used, while we might not be able to facilitate this kind of functionality directly I think the general interface they present is more robust.

File: src/main/java/io/nitric/api/queue/Queue.java [hunk 2]
```diff
@@ -53,50 +52,59 @@
...
+    public String getName() {
+        return name;
+    }
+
     /**
      * Send the given task to the client queue.
      *
      * @param task the task to send to the queue (required)
      */
     public void send(Task task) {
```
Comment (tjholm) [inline src/main/java/io/nitric/api/queue/Queue.java:93]: Might be good to consider rolling send and sendBatch together, then we can look at deprecating and removing the singular method on the membrane side. We can still present the singular method SDK side, but have it utilise the batch implementation.

File: src/main/java/io/nitric/proto/event/v1/EventServiceGrpc.java [hunk 3]
```diff
@@ -287,11 +287,11 @@ public void invoke(Req request, io.grpc.stub.StreamObserver<Resp> responseObserv
...
-      synchronized (EventGrpc.class) {
+      synchronized (EventServiceGrpc.class) {
         result = serviceDescriptor;
         if (result == null) {
           serviceDescriptor = result = io.grpc.ServiceDescriptor.newBuilder(SERVICE_NAME)
-              .setSchemaDescriptor(new EventFileDescriptorSupplier())
+              .setSchemaDescriptor(new EventServiceFileDescriptorSupplier())
               .addMethod(getPublishMethod())
               .build();
         }
```
Comment (tjholm) [inline src/main/java/io/nitric/proto/event/v1/EventServiceGrpc.java:297]: Is it feasible to also not version generated proto files to reduce noise as well? In some runtimes its not possible e.g. Go/PHP. But for runtimes that use packaging prior to deploy.

File: pom.xml [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
     <groupId>io.nitric</groupId>
     <artifactId>java-sdk</artifactId>
-    <version>0.2.0</version>
+    <version>0.5.0</version>
```
[the remark this answers is not in the document]
Comment (medgar-nitric) [reply pom.xml:7]: I have some existing 0.3.0-SNAPSHOT and 0.4.0-SNAPSHOT deployments on the staging repository when doing templates testing:  https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/
I cant delete these files. Keeping the history ordering I think its more appropriate to bump the version past those versions.

File: src/main/java/io/nitric/api/document/CollectionDocRef.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ */
+
+import io.nitric.util.Contracts;
+
+import java.util.Map;
+
+/**
+ * Provides a collection Document Reference class.
+ */
+public class CollectionDocRef<T> extends AbstractDocRef<T> {
```
[the remark this answers is not in the document]
Comment (medgar-nitric) [reply src/main/java/io/nitric/api/document/CollectionDocRef.java:30]: Sure no problem will change to DocumentRef and rationalize class hierarchy and rely on runtime errors for invalid subcollection operations.

File: src/main/java/io/nitric/api/document/SubCollection.java (hunk 1 above)
Comment (medgar-nitric) [reply src/main/java/io/nitric/api/document/SubCollection.java:73]: This check is to ensure a subcollection document reference has a valid parent path hierarchy. For example below:

```
var doc = Documents.collection("customers").collection("orders").doc("123")

/collections/<unknown>/orders/123
```

File: src/main/java/io/nitric/api/queue/Queue.java (hunk 2 above)
Comment (medgar-nitric) [reply src/main/java/io/nitric/api/queue/Queue.java:93]: Yeah, that makes sense. 

The ```sendBatch``` method is also good in terms of signaling failed send, we should probably add this to the signature of the ```sendTask``` method.

```
public FailedTask sendTask(Task)
```

File: src/main/java/io/nitric/proto/event/v1/EventServiceGrpc.java (hunk 3 above)
Comment (medgar-nitric) [reply src/main/java/io/nitric/proto/event/v1/EventServiceGrpc.java:297]: I will look into it.

File: pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (tjholm) [reply pom.xml:7]: no worries, if the deployments already exist thats fine.

**Review by tjholm — APPROVED** [review]

**Pull request merged**